modules/telemetry/src/metric_record.cpp

```diff
@@ -0,0 +1,152 @@
+
+namespace internal {
+namespace {
+[[nodiscard]] auto stringToInt64(const std::string& str) -> std::optional<int64_t> {
+  const char* start = str.c_str();
+  char* end{};
+  errno = 0;
+
+  static constexpr int BASE = 10;
+  int64_t result = std::strtoll(start, &end, BASE);
+
+  if (errno != ERANGE && *end == '\0') {
+    return result;
+  }
+
+  return std::nullopt;
+}
+
+[[nodiscard]] auto jsonToValue(const nlohmann::json& json_value) -> std::optional<Metric::ValueType> {
+  if (json_value.is_boolean()) {
+    return json_value.get<bool>();
+  }
+
+  if (json_value.is_number_float()) {
+    return json_value.get<double>();
+  }
+
+  if (json_value.is_number_integer() || json_value.is_number_unsigned()) {
+    return json_value.get<int64_t>();
+  }
+
+  if (json_value.is_string()) {
+    auto value_str = json_value.get<std::string>();
+    // According to JSON specification integer bigger than 32bits are converted to string, so we need to check
+    // if the string is actually a number.
+    if (const auto value_i = stringToInt64(value_str); value_i.has_value()) {
+      return value_i.value();
+    }
+
+    return value_str;
+  }
+
+  return {};
+}
+
+void jsonToValues(const nlohmann::json& json, std::unordered_map<std::string, Metric::ValueType>& values,
+                  std::string_view key_prefix) {
+  for (const auto& [key, value] : json.items()) {
+    const auto full_key = key_prefix.empty() ? key : fmt::format("{}.{}", key_prefix, key);
+    if (const auto value_opt = jsonToValue(value); value_opt.has_value()) {
+      values[full_key] = value_opt.value();
+    } else if (value.is_object()) {
+      // If it's an object we need to recursively call this function.
+      // The key_prefix is updated to include the current key.
+      // For a nested value the key will be the concatenation of all the keys in the path.
+      jsonToValues(value, values, full_key);
+    } else {
+      // NOTE: we do not support arrays.
+      LOG(ERROR) << fmt::format("Failed to parse value for key: {}, value: {}", full_key, value.dump());
+    }
+  }
+}
+
+}  // namespace
+
+auto jsonToValuesMap(std::string_view json) -> std::unordered_map<std::string, Metric::ValueType> {
+  const auto json_obj = nlohmann::json::parse(json);
+  std::unordered_map<std::string, Metric::ValueType> values;
+
+  jsonToValues(json_obj, values, "");
+
+  return values;
+}
+}  // namespace internal
+
+class MetricRecorder {
+public:
+  explicit MetricRecorder();
+  static void registerMetricSink(std::unique_ptr<IMetricSink> sink);
+
+  static void record(const Metric& metric);
+
+private:
+  [[nodiscard]] static auto instance() -> MetricRecorder&;
+  void processMeasureEntries(const Metric& entry);
+
+private:
+  std::mutex sink_mutex_;
+  std::vector<std::unique_ptr<IMetricSink>> sinks_ ABSL_GUARDED_BY(sink_mutex_);
+  concurrency::MessageQueueConsumer<Metric> measure_entries_consumer_;
+};
+
+void registerMetricSink(std::unique_ptr<IMetricSink> sink) {
```

Comment:
also here I would call it just registerSink and overload on the sink type.
IMO
```
registerSink(myMetricSink);
registerSink(myLogSink);
```
is cleaner

Reply:
I am not sure, I think it is important to explicitly state that the sink for metrics and log are different.

Reply:
I think since they have different interface types its clear, but we can also have more explicit names here.
Better write 5 letters more than not being expressive